driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCursor.java

```diff
@@ -25,67 +26,82 @@
 import org.reactivestreams.Subscription;
 
 import java.util.NoSuchElementException;
+import java.util.concurrent.BlockingDeque;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingDeque;
 import java.util.concurrent.TimeUnit;
 
-public class SyncMongoCursor<T> implements MongoCursor<T> {
+class SyncMongoCursor<T> implements MongoCursor<T> {
+    private static final Object COMPLETED = new Object();
+
     private volatile Subscription subscription;
-    private volatile CountDownLatch latch;
     private volatile T next;
-    private volatile Throwable error;
-    private volatile boolean complete;
+    private final BlockingDeque<Object> results = new LinkedBlockingDeque<>();
```

Comment:
Now adding all results, error and completed flag to the same queue so they are delivered in order to next() method.